### DISCUSSION TITLE
Column alignment failure using marginalia--fields

### DISCUSSION THREAD

**Discussion by mgsb** [minad/marginalia#179] [Q&A] [answered]
While using `marginalia--fields` in an annotator function, column alignment
fails, most likely due to pilot error. The face attributes work correctly.
This is for a mode that includes installing MicroPython (non-emacs) packages 
so the default annotator can't be used. The relevant functions:

``` emacs-lisp
(defun mp--package-annotator (pkg-list pkg-name)
  "Annotate the candidate PKG-NAME using metadata in PKG-LIST."
  (let* ((pkg (assoc (intern pkg-name) pkg-list))
         (description (cdr (assoc 'description pkg)))
         (version (cdr (assoc 'version pkg))))
    (if (fboundp 'marginalia--fields)
        (marginalia--fields
         (version :face 'marginalia-version)
         (description :face 'marginalia-documentation))
      (format "    %s - %s" version description))))

(defun mp-install-package ()
  "Query for one or more MicroPython packages to install using `mip install`."
  (interactive)
  (when (boundp 'marginalia-command-categories)
    (unless (assoc (intern  "mp-install-package") marginalia-command-categories)
      (add-to-list 'marginalia-command-categories
                   '(mp-install-package . not-emacs-package))))
  (let* ((pkg-list (mp--fetch-index))
         (completion-extra-properties `(:annotation-function ,(apply-partially 'mp--package-annotator pkg-list)))
         (pkgs (completing-read-multiple "Package: " pkg-list)))
    (message (format "installing: %s ..." (string-join pkgs " ")))
    (mp--install pkgs t)))
```
Any suggestions/corrections appreciated.

**Comment by minad** [comment, marked as answer]
Please note that [Marginalia is a user facing package and is not intended to be used as a library](https://github.com/minad/marginalia?tab=readme-ov-file#adding-custom-annotators-or-classifiers).

**Comment by mgsb** [comment]
Ok, thanks for the clarification. I misunderstood the previous version of the README when it referenced `marginalia--fields` and annotation functions. I re-wrote my annotator to directly pad the spacing and use faces.

Perhaps it would help to use the `annotation-function` first in `marginalia-cycle` so that developers supplying a function would not be surprised to not see their function output being used when their prompt matches a marginalia annotator.

Also, maybe the annotator regex's should be tighter so developers could use a prompt like "Python package: " and not trigger the marginalia default.

I used the following to insure my `annotation-function` will be used. Does this look right?

```emacs-lisp
(when (require 'marginalia nil t)
  (add-to-list 'marginalia-command-categories
               '(mp-install-package . not-emacs-package)))
```

**Reply by minad** [reply]
> Perhaps it would help to use the annotation-function first in marginalia-cycle so that developers supplying a function would not be surprised to not see their function output being used when their prompt matches a marginalia annotator.

No, Marginalia intends to override builtin annotators by default. In your case you can prevent problems by specifying both a `category` (e.g. `python-package`) and an `annotation-function`. Then your annotation function won't be overridden. Specifying the category is a good idea anyway, since it will enable Embark support and will make it possible to use overrides in `completion-category-overrides`.

> I used the following to insure my annotation-function will be used. Does this look right?

No, this won't be necessary then. Marginalia is designed such that you don't have to mention it at all in your package. You can work completely independently. Otherwise our recommendation in https://github.com/minad/marginalia?tab=readme-ov-file#adding-custom-annotators-or-classifiers would be a bit odd.

**Reply by mgsb** [reply]
Ok, I re-read the [Programmed Completion](https://www.gnu.org/software/emacs/manual/html_node/elisp/Programmed-Completion.html) section in the manual and then found this [article](https://kisaragi-hiu.com/emacs-completion-metadata/) (written in the context of Marginalia and Embark) that helped clear some of the fog for me. I used a variant of `k//mark-category` from the article as my collection function in `completing-read`, with a custom category. Now things work without any references to Marginalia. I currently do not have any ideas for functions to use with Embark, but will keep it in mind for future use.

I appreciate your packages and support - thanks!

**Reply by minad** [reply]
Just a small nit - `k//mark-category` is not written correctly. Unfortunately there is a lot of incomplete code in blog posts.  In case of doubt, look at the completion tables in Emacs code or well established packages like Embark. Embark comes with the following helper (how `k//mark-category` should have actually be written):

~~~ elisp
(defun embark--with-category (category candidates)
  "Return completion table for CANDIDATES of CATEGORY with sorting disabled."
  (lambda (string predicate action)
    (if (eq action 'metadata)
        `(metadata (category . ,category))
      (complete-with-action
       action candidates string predicate))))
~~~

**Reply by mgsb** [reply]
Thanks for the correction! Reading the manual (again), it's not clear why/when it's needed to specify the sort functions. Searching through installed packages, I see a variety of usages. Can you explain? I looked at minibuffer.el and not sure I get all what's happening. Also, how does this interact with Vertico (or not)?

**Reply by minad** [reply]
It is not about the sort function, but about `complete-with-action`. I've removed the sort function from above. Hope it is more clear now.

**Reply by mgsb** [reply]
Sorry - I understood about `complete-with-action`.
